File: bilfind_app/lib/models/conversation_model.dart
import 'package:bilfind_app/constants/app_constants.dart';
import 'package:bilfind_app/models/message_model.dart';
import 'package:bilfind_app/models/post_model.dart';

import 'program.dart';

class ConversationModel {
  ConversationModel({
    required this.id,
    required this.post,
    required this.ownerPhoto,
    required this.ownerName,
    required this.ownerEmail,
    required this.ownerId,
    required this.ownerDepartment,
    required this.senderPhoto,
    required this.senderName,
    required this.senderEmail,
    required this.senderId,
    required this.senderDepartment,
    required this.createdAt,
    required this.status,
    required this.messages,
  });
  late final String id;
  late final PostModel post;
  late final String? ownerPhoto;
  late final String ownerName;
  late final String ownerEmail;
  late final String ownerId;
  late final String ownerDepartment;
  late final String? senderPhoto;
  late final String senderName;
  late final String senderEmail;
  late final String senderId;
  late final String senderDepartment;
  late final DateTime createdAt;
  late final String status;
  late final List<MessageModel> messages;

  ConversationModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    post = PostModel.fromJson(json['post']);
    ownerPhoto = json['ownerPhoto'];
    ownerName = json['ownerName'];
    ownerEmail = json['ownerEmail'];
    ownerId = json['ownerId'];
    ownerDepartment = json['ownerDepartment'];
    senderPhoto = json['senderPhoto'];
    senderName = json['senderName'];
    senderEmail = json['senderEmail'];
    senderId = json['senderId'];
    senderDepartment = json['senderDepartment'];
    createdAt = DateTime.parse(json['createdAt']).toLocal();
    status = json['status'];
    messages = (json['messages'] as List)
        .map((e) => MessageModel.fromJson(e))
        .toList();
  }

  String getChatImageUrl() {
    String? imageUrl;

    if (senderId == Program().userModel?.id) {
      imageUrl = ownerPhoto;
    } else {
      imageUrl = senderPhoto;
    }

    imageUrl ??= AppConstants.defaultProfilePhoto;

    return imageUrl;
  }

  String getChatUserName() {
    if (senderId == Program().userModel?.id) {
      return ownerName;
    }

    return senderName;
  }
}

File: backend/src/models/conversation_model.ts
import { Expose, Transform, Type } from "class-transformer";
import { ObjectId } from "mongodb";
import { MessageModel } from "./message-model";
import { IsEnum } from "class-validator";
import { User, UserResponseDTO } from "./user-model";
import { PostResponseDTO } from "./post-model";
import { Mapper } from "../utils/mapper";
import { Departments } from "../utils/enums";

export enum ConversationStatus {
  WAITING = "WAITING",
  ACTIVE = "ACTIVE",
  DELETED = "DELETED",
}

export class ConversationModel {
  @Transform((value) => value.obj._id.toString())
  @Expose()
  _id?: ObjectId;

  @Expose()
  postId: string;

  @Expose()
  senderUserId: string;

  @Expose()
  postOwnerUserId: string;

  @Expose()
  createdAt: Date;

  @Expose()
  @IsEnum(ConversationStatus)
  status: ConversationStatus;

  @Expose()
  @Type(() => MessageModel)
  messages: MessageModel[];
}

export class ConversationResponseDto {
  @Transform((value) => value.obj._id.toString())
  @Expose()
  id: ObjectId;

  @Expose()
  post: PostResponseDTO;

  //owner info
  @Expose()
  ownerPhoto?: string;

  @Expose()
  ownerName: string;

  @Expose()
  ownerEmail: string;

  @Expose()
  ownerId: string;

  @Expose()
  ownerDepartment: Departments;

  //sender info
  @Expose()
  senderPhoto?: string;

  @Expose()
  senderName: string;

  @Expose()
  senderEmail: string;

  @Expose()
  senderId: string;

  @Expose()
  senderDepartment: Departments;

  @Expose()
  createdAt: Date;

  @Expose()
  @IsEnum(ConversationStatus)
  status: ConversationStatus;

  @Expose()
  @Type(() => MessageModel)
  messages: MessageModel[];
}

export const mapToConversationResponseDTO = (
  conversation: ConversationModel,
  post: PostResponseDTO,
  senderUser: User,
  postOwnerUser: User
): ConversationResponseDto => {
  return Mapper.map(ConversationResponseDto, {
    ...conversation,
    id: conversation._id!.toString(),
    post,
    //post owner
    ownerPhoto: postOwnerUser.profilePhoto,
    ownerName: postOwnerUser.name + " " + postOwnerUser.familyName,
    ownerDepartment: postOwnerUser.departmant,
    ownerEmail: postOwnerUser.email,
    ownerId: postOwnerUser._id!.toString(),
    //sender
    senderPhoto: senderUser.profilePhoto,
    senderName: senderUser.name + " " + senderUser.familyName,
    senderDepartment: senderUser.departmant,
    senderEmail: senderUser.email,
    senderId: senderUser._id!.toString(),
  });
};
